### PR TITLE
VideoBackends: Set the maximum range when the depth range is oversized.

### PR DESCRIPTION
**Pull request by Armada651**
This PR fixes regressions introduced in PR #4471 and PR #4085. It also contains a code cleanup that removes a hack for OpenGL ES, we should consider always relying on slow depth when that API is in use.

### Regression 1

The first regression is very simple, during PR #4471 I worked with the assumption that an oversized depth range would always result in a 0:1 depth range in the viewport. However in some games this may not be the case, there are situations where the far value is oversized, but the range value is normal, thus we get a depth range that is smaller than 0:1.

This PR fixes that simply by ensuring the range is set properly when one of the parameters is oversized.

### Regression 2

The second regression is more complicated, it's an interplay between PR #4471 and PR #4085. PR #4085 introduced depth clamping together with user-defined clipping planes to fix clipping issues in games such as Sonic Adventures. However depth clamping will not only clamp the viewport transform to the depth range as we intended, but it will also clamp depth values outputted by the pixel shader to the depth range.

This is can be a problem in games such as the PAL50 version of Mario Kart, which uses a ztexture to clear the depth buffer outside of the depth range. Before PR #4471 this wasn't a problem, because PR #4085 made sure the depth range was always set to the maximum value. To fix this we simply disable depth clamping while a ztexture is enabled.

### Bonus: glDepthRangedNV support

Because no one likes hacks upon hacks I decided this would be a good time to introduce support for the `NV_depth_buffer_float` extension in OpenGL. The benefit of this extension is that it allows OpenGL to handle oversized depth ranges without any hacks just like the console can. When this extension is supported the OpenGL backend will always let OpenGL process the depth range instead of trying to handle it in the vertex shader, which can lead to z-fighting issues.

### CODE REVIEW

**Comment by Armada651** [comment]
`simpsons-game` and `mii-channel` are bugged, but this seems to be a mesa driver issue as I can't reproduce it on nvidia.

**Comment by GitToTheHub** [comment]
If that should fix [Issue 10073](https://bugs.dolphin-emu.org/issues/10073) the error does still occure in OGL and D3D. I have still a black flickering in Mario Kart Wii.

**Comment by Armada651** [comment]
@GitToTheHub Can you capture a FIFO log so I can take a look at it?

**Comment by GitToTheHub** [comment]
How much frames do you need?

**Comment by Armada651** [comment]
Just a few to see the flicker, so something like 5 frames.

**Comment by GitToTheHub** [comment]
I don't know why i can't upload the file to github.. I uploaded it as 7z to an external hoster. I recorded 5 frames: http://www.filedropper.com/mariokartwiiblackflickering5frames

**Comment by Armada651** [comment]
@GitToTheHub I can see the flickering, but it doesn't seem related to this PR at all.

**Comment by GitToTheHub** [comment]
ok i thought it would be related to pr #4471 and should be fixed with that pr. But maybe i missunderstood that :)

**Comment by GitToTheHub** [comment]
I rechecked it. I don't have a flicker in [Build 5.0-1600](http://dl.dolphin-emu.org/builds/dolphin-master-5.0-1600-x64.7z).  In [Build 5.0-1605](http://dl.dolphin-emu.org/builds/dolphin-master-5.0-1605-x64.7z) i have the flickering which was created by the pr #4471.

**Comment by Armada651** [comment]
@GitToTheHub I checked it again and you are right that PR #4471 caused a regression for that game, sorry about that.

The problem is, this game doesn't have an oversized depth range. It breaks when using the standard depth range handling in the graphics drivers, which should be more accurate than our own depth range algorithm, not less accurate.

**Comment by Armada651** [comment]
I've found the problem, this has something to do with using our own depth clipping planes. I'm not yet sure how I'm going to fix that.

**Comment by Armada651** [comment]
@GitToTheHub I've added a fix for the OpenGL and Vulkan backends, but it will be difficult to fix this in the D3D backend.

**Comment by JMC47** [comment]
If it's not possible to fix this in D3D11/12, I'm for just fixing this in OpenGL/Vulkan for now.

It's one game (PAL version at that...) and I'm fairly certain PAL60 would avoid the issue as well.  Adding a ton of code complexity/fallbacks to fix it seems like overkill.

**Comment by GitToTheHub** [comment]
@Armada651 Thanks for looking deeper in this issues. Sounds not so good, that it maybe cannot be fixed in D3D. Im using a default European PAL Version. Should that not be supported?
In Mario Kart Double Dash (PAL-Version) i have in D3D the issue, that the screen in the menu stays black. I can't see anything. I have not tested other PAL-Games yet, but i fear, that no PAL-Games will work anymore with D3D. Can this be right?

**Comment by Armada651** [comment]
> I'm using a default European PAL Version. Should that not be supported?

You should see if there is an option to run the game at 60Hz. The flickering is caused by Mario Kart rendering more frames than it is displaying at 50Hz. 

> In Mario Kart Double Dash (PAL-Version) i have in D3D the issue, that the screen in the menu stays black.

That sounds like a separate issue. Again, if you submit FIFO logs on the issue tracker then I can look into it.

**Comment by degasus** [comment]
Mesa doesn't support NV_depth_buffer_float :/

**Comment by Armada651** [comment]
@degasus That's not really a problem, if NV_depth_buffer_float is not supported we'll just handle the oversized depth range in the vertex shader. It's just less clean that's all.

**Comment by GitToTheHub** [comment]
@Armada651
Hi, i tested a few pal games with the current PR and the most games has no issues. But some issues i found anyway:
- Smash Bros. Brawl is broken. The charakters are not visible anymore and the screen is not completly visible.
- Super Smash Bros. Melee has a few issues. The screen is really small and the shadows has issues.
- Bomberman Generation: There was only 3/4 of the screen visible, if you play the story mode, the shadows are there misplaced.

I retested Mario Kart Double Dash, and sorry now, i have no issues, there is no black screen. I don't know, why i had one time this issue. The game is normally working.

But Mario Kart Wii (PAL) has now greater issues. There is now really a black screen in D3D, here the fifo log: https://ufile.io/a5b56. Though, if you click through the black screen to the game, the flicker is gone now. But, the transparent blue box in the upper left corner is visible again.

**Comment by Armada651** [comment]
I assume this is all on D3D? If so, I must've made an error in that backend.

**Comment by GitToTheHub** [comment]
Right this is all in D3D. I will test Mario Kart Wii in OGL.

**Comment by Armada651** [comment]
Found the problem, it's a bit embarrassing. Thank you for finding it early.

**Comment by GitToTheHub** [comment]
Ok can happen. This would be great, if the parentheses would fix the problem :)
In OGL i have found no issues except in Mario Kart Wii (PAL, 50Hz) there is still the flickering :(

**Comment by Armada651** [comment]
That could be a driver issue, which GPU are you using?

**Comment by GitToTheHub** [comment]
Im using the AMD Radeon HD 5870. It uses OGL 3.2. If it helps, here the fifo log: https://ufile.io/d881b

**Comment by JMC47** [comment]
That doesn't support the necessary extension.  You'll have to turn on 60Hz mode in MKWii to avoid the flickering.

**Comment by Armada651** [comment]
@JMC47 That's not true actually, glDepthRangedNV is not required to render Mario Kart Wii PAL50 correctly.

But it's possible AMD behaves differently than Nvidia. It's very easy to make a mistake about whether pixel shader depth output should be clamped or not.

**Comment by JMC47** [comment]
Oh boy.  I'll test on my AMD later with the fifolog.

**Comment by GitToTheHub** [comment]
@JMC47 
Thanks for the hint. I enabled in the dolphin settings PAL60-Mode and the flickering is gone in OGL. But can this broke other wii games? Can this option enabled via the game ini for users who don't know how to fix that issue?

**Comment by JMC47** [comment]
PAL60 doesn't break any games.

**Comment by GitToTheHub** [comment]
And why is this not enabled be default if it does not break anything? Is 60Hz not better as 50Hz?

**Comment by JMC47** [comment]
@GitToTheHub it should work with PAL50 again.

PAL60 is usually considered better, I don't know why it's not the default.

**Comment by GitToTheHub** [comment]
Hi, i wanted to test the actual pr-build, but i have no access anymore to https://dl.dolphin-emu.org/prs. A 403 Error "Forbidden" is returned.

**Comment by Armada651** [comment]
@GitToTheHub At the bottom of this PR press "Show all checks" then click on "Details" next to the pr-win-x64 bot. There you can find the link to the 7z file.

**Comment by GitToTheHub** [comment]
@Armada651 Ok thank you, i found it :) I will test it later. Thanks.

**Comment by GitToTheHub** [comment]
@Armada651 Thanks for the fix. Mario Kart Wii does now work in 50Hz and 60Hz mode, in D3D and OGL. All games which before has issues are fixed now. But i tested now other wii games and they have issues with D3D though. In OGL are no issues.
Here are the Games. It does not matter if running them in 50Hz or 60Hz:
- New Super Mario Bros.: No Title is shown when the game starts. Only the Characters are running. Fifo Log: https://ufile.io/5fa7d1
- Bomberman Blast: Black Screen but game is running
- Mario Strikers: Menu Text is missing as in New Super Mario Bros.
- In Super Mario Galaxy 1 & 2 are missing Text issues too
Mario Kart Wii and Super Smash Bros. Brawl are running fine in D3D. Maybe D3D is really a bad backend and only OGL should be supported :)

**Comment by GitToTheHub** [comment]
I made a speed test vs OGL and D3D in Mario Kart Wii and D3D is 60FPS faster as OGL. So without D3D you have a speed loss :/

**Comment by JMC47** [comment]
Well you're on an AMD.  It's not our fault their drivers are slow as hell at OpenGL.

**Comment by GitToTheHub** [comment]
hmm, ok thats bad. Is Nvidia faster in OGL?

**Comment by JMC47** [comment]
Yep

**Comment by Armada651** [comment]
@GitToTheHub I tried your Fifo log, but the title shows up fine for me on D3D and OGL. I'm on an Nvidia card though, maybe @JMC47 can try to confirm it on an AMD?

**Comment by JMC47** [comment]
I can check it out on non-OpenGL pretty easily

**Comment by Armada651** [comment]
@GitToTheHub Also can you confirm whether this issue also happens on OGL? And does it happen on the latest development version?

**Comment by GitToTheHub** [comment]
@Armada651 What you don't have the issue with D3D? In OGL is everything fine here. What you mean with the "latest development version"?

**Comment by GitToTheHub** [comment]
I think i know what you mean. I tried Build 5.0-2682 and the issue is there not. Neither in D3D or OGL. But the Mario Kart Wii issues are there.

**Comment by Armada651** [comment]
@GitToTheHub I don't have the issue with D3D on the build from this PR.

**Comment by JMC47** [comment]
@Armada651 I checked his fifolog, works on the r9 285 on D3D11, D3D12 and Vulkan.  NSMBWii works, Mario Kart Wii works.

**Comment by GitToTheHub** [comment]
Ok i found the problem. This has something to do if i test different dolphin versions and switch the backends. There is then a faulty dolphin cache or so. If i clear the cache folder the problems are gone. In Super Mario Galaxy 2, for example, the game has start with a blue screen and didn't show the message, that no controllers are connected. If i clear the cache, the game is working normally. So the changes are all ok. Sorry for that.

**Comment by GitToTheHub** [comment]
Could the cache maybe generally cleared,if you run a different dolphin build?

**Comment by Armada651** [comment]
Shaders don't change every build, but we do keep a separate version number for the shaders and the cache is cleared when the shaders change significantly.

**Comment by GitToTheHub** [comment]
Ok, but this machanism has not work this time, or? I don't know why i must clear the cache dir, so the problems are gone. But ok, the problem is solved :)

**Comment by GitToTheHub** [comment]
Hey, i found the shader bug. Here the steps to reproduce the bug. I tested it with Super Mario Galaxy 2:
- Start with this pr build Super Mariox Galaxy 2 in D3D and close the emulation after the first screen.
- Close Dolphin
- Start the latest development build, set backend to OGL and start Super Mario Galaxy 2. Close the emulation.
- Set now the backend to D3D and start Super Mario Galaxy 2 again. Now you have a blue screen on the first screen.
The D3D error will not be reolved if you restart Dolphin. You must start another Dolphin Version such as this pr and must start the game directly in D3D to resolve this issue.

Maybe this problem is not related to this pr and someone can open a ticket in the issue tracker.

**Comment by GitToTheHub** [comment]
I tested the same steps mit New Super Mario Bros. and guess what, the text is gone :)
smnp01-3

**Comment by stenzek** [comment]
@GitToTheHub the shader cache is tied to the git commit that the build was based on, see: https://github.com/dolphin-emu/dolphin/blob/master/Source/Core/Common/LinearDiskCache.h#L180

However, since PR's are built by taking master and applying the branch/PR/patch on top, the git revision remains the same _(as far as I'm aware this is the case, if it's not, please correct me)_.

Therefore, if the development version you used was the same as the one that the PR was built against, you may run into issues as Dolphin will re-use the existing cache, and you'll have a combination of "old" and "new" shaders (or "old" shaders using the "new" behavior).

Not sure if that's what's going on here, but it may be a possible explanation, in which case everything is working as intended. If you are able to reproduce this problem with two different **development** builds, then that is a concern.

**Comment by GitToTheHub** [comment]
@stenzek 
Hi, i understand what you mean. If i play a game only in D3D and switch between the latest development version and this PR, there are no problems. Only If i change between a switch the backend first to OGL and later to D3D the problems will occure. Do you have two different development builds for me where the shader cache should be cleared, so that i can test it?

**Comment by GitToTheHub** [comment]
I tested with two different development Builds: 5.0-2577 and 5.0-2684 and the bug does still here occure.

**Comment by stenzek** [comment]
@GitToTheHub Interesting. Switching the backends shouldn't have anything to do with the caches, they should be completely separate. I did test briefly and couldn't reproduce the problem, but if I have time tomorrow I'll try to look into it some more.

Feel free to open an issue on the tracker - it is not related to this PR.

**Comment by GitToTheHub** [comment]
You couldn't reproduce it? Its important, that you first start the game with D3D 11, and in another Dolphin Version then first with OGL and a second time with D3D 11. I can produce the error all the time. If i have time i will open a ticket.

File: Source/Core/VideoCommon/RenderBase.cpp [hunk 1]
```diff
@@ -940,3 +940,30 @@ void Renderer::DumpFrameToImage(const FrameDumpConfig& config)
...
+bool Renderer::UseVertexDepthRange() const
+{
+  // We can't compute the depth range in the vertex shader if we don't support depth clamp.
+  if (!g_ActiveConfig.backend_info.bSupportsDepthClamp)
+    return false;
+
+  if (g_ActiveConfig.backend_info.bSupportsOversizedDepthRanges)
+  {
+    // We support oversized depth ranges, but we need a full depth range if a ztexture is used.
+    return bpmem.ztex2.type != ZTEXTURE_DISABLE;
```
Comment (degasus) [inline Source/Core/VideoCommon/RenderBase.cpp]: I miss an early_z check here. As it disables the ztexture.

**Comment by degasus** [comment]
small nitpick, else LGTM

**Pull request merged**